ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/objectstore/slot/DiskSlots.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2020, Red Hat, Inc. and/or its affiliates,
+ * and individual contributors as indicated by the @author tags.
+ * See the copyright.txt in the distribution for a
+ * full listing of individual contributors.
+ * This copyrighted material is made available to anyone wishing to use,
+ * modify, copy, or redistribute it subject to the terms and conditions
+ * of the GNU Lesser General Public License, v. 2.1.
+ * This program is distributed in the hope that it will be useful, but WITHOUT A
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ * PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
+ * You should have received a copy of the GNU Lesser General Public License,
+ * v.2.1 along with this distribution; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA  02110-1301, USA.
+ *
+ * (C) 2020,
+ * @author Jonathan Halliday (jonathan.halliday@redhat.com)
+ */
+package com.arjuna.ats.internal.arjuna.objectstore.slot;
+
+import com.arjuna.ats.arjuna.logging.tsLogger;
+
+import java.io.*;
+import java.util.zip.CRC32;
+
+/**
+ * File backed implementation of the SlotStore backend.
+ * WARNING: this is currently a test implementation and is NOT SAFE for production use yet.
+ *
+ * @author Jonathan Halliday (jonathan.halliday@redhat.com), 2020-04
+ */
+public class DiskSlots implements BackingSlots {
+
+    private SlotStoreEnvironmentBean config;
+    private RandomAccessFile[] slots;
+
+    @Override
+    public synchronized void init(SlotStoreEnvironmentBean config) throws IOException {
+        if (this.config != null) {
+            throw new IllegalStateException("already initialized");
+        }
+
+        this.config = config;
+
+        File storeDir = new File(config.getStoreDir());
+        if (!storeDir.exists() && !storeDir.mkdirs()) {
+            throw new IOException(tsLogger.i18NLogger.get_dir_create_failed(storeDir.getCanonicalPath()));
+        }
+        slots = new RandomAccessFile[config.getNumberOfSlots()];
```

Comment:
It's not memory mapped, but your point is still relevant.  The Java API is cursor based - you seek to the location you want to write, then write it. So you'd get lock contention on the shared cursor. The fd.sync is also a problem - it syncs the whole file. The O/S will be dirty tracking at the block level, but will still incur some overhead to find the dirty blocks to flush. So, you'd certain get less resource (RAM) usage, but I don't see a performance improvement as likely. It's more the other way around I think.

However...

The pmem backend is designed to use memory mapping over a single file. It's also JDK14 only, because that's the first version that has the pmem support in the JVM. So it can and does use the newer Java API that allows both absolute position writes (no cursor to lock, you specify the offset as a call parameter instead) and region flush - calling the new force(offset, length) instead of force().  Although it's designed for pmem, it should be possible to run almost the same code in non-pmem mode, at which point it's equivalent to the single file version of the store you suggest. So another reason the DiskSlots one is multi-file, is to provide an alternative for comparison.  Finally, the multi-file approach is more robust to slotSize changes, as it can grow files. With a single file, you have to delete and recreate the whole thing if you want to change the slotSize.

Reply:
Good explanation.